rewrite-java/src/main/java/org/openrewrite/java/UsePrecompiledRegExpForReplaceAll.java

```diff
@@ -0,0 +1,119 @@
+        private static final JavaTemplate MATCHER_TEMPLATE = JavaTemplate
+                .builder("#{any()}.matcher( #{any(java.lang.CharSequence)}).replaceAll( #{any(java.lang.String)})")
+                .contextSensitive()
+                .imports("java.util.regex.Pattern")
+                .build();
+        private static final MethodMatcher REPLACE_ALL_METHOD_MATCHER = new MethodMatcher("java.lang.String replaceAll(String, String)");
+
+        @Override
+        public J.Block visitBlock(final J.Block block, final ExecutionContext executionContext) {
+            final J.Block processedBlock = super.visitBlock(block, executionContext);
+            final Cursor parentOrThrow = getCursor().getParentOrThrow();
+            if (parentOrThrow.getValue() instanceof J.ClassDeclaration) {
+                final Object regExp = parentOrThrow.pollNearestMessage(REG_EXP_KEY);
+                if (regExp == null) {
+                    return processedBlock;
+                }
+                patternProcessingInProgress.set(false);
+
+                final String regExpName = parentOrThrow.pollNearestMessage(REG_EXP_KEY + "_NAME");
+                return super.visitBlock(
+                        JavaTemplate
+                                .builder("private static final java.util.regex.Pattern " + regExpName + " = Pattern.compile(#{});")
+                                .contextSensitive()
+                                .imports("java.util.regex.Pattern")
+                                .build()
+                                .apply(
+                                        getCursor(),
+                                        processedBlock.getCoordinates().firstStatement(),
+                                        regExp),
+                        executionContext);
+            }
+            return processedBlock;
+        }
+
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext executionContext) {
+            final boolean alreadyProcessingAPattern = patternProcessingInProgress.get();
+            if (!REPLACE_ALL_METHOD_MATCHER.matches(method) || alreadyProcessingAPattern) {
+                return super.visitMethodInvocation(method, executionContext);
+            }
+
+            patternProcessingInProgress.set(true);
+
+            final J.MethodInvocation invocation = super.visitMethodInvocation(method, executionContext);
+            final String regexPatternVariableName = VariableNameUtils.generateVariableName(PATTERN_VAR, getCursor(), VariableNameUtils.GenerationStrategy.INCREMENT_NUMBER);
+            final Object regexPatternIdentifier = new J.Identifier(
+                    randomId(),
+                    Space.SINGLE_SPACE,
+                    Markers.EMPTY,
+                    regexPatternVariableName,
+                    JavaType.buildType(Pattern.class.getName()),
+                    null);
+
+            final Cursor cursor = getCursor();
```

Comment:
Is it safe to declare a variable for the cursor and reuse it or should I always call the getCursor() function?

Reply:
Declaring a variable is safe. But in case you were to modify the input object (in this case `method`) then you have to create a new `Cursor` to point to that updated instance, if you later want to modify the instance again using a `JavaTemplate`. To this end you can use `updateCursor(Object)`.

Here you do in fact do this on line 98 with the `super.visitMethodInvocation()` call. So I suggest the following:
```suggestion
            final Cursor cursor = updateCursor(invocation);
```